File: app/src/main/java/ua/hospes/nfs/marathon/ui/race/widgets/DriverTimeView.java
package ua.hospes.nfs.marathon.ui.race.widgets;

import android.content.Context;
import android.util.AttributeSet;

import ua.hospes.nfs.marathon.domain.sessions.models.Session;
import ua.hospes.nfs.marathon.utils.TimeUtils;

/**
 * @author Andrew Khloponin
 */
public class DriverTimeView extends TimeView {
    private Session session;
    private long prevDuration = 0L;


    public DriverTimeView(Context context) {
        super(context);
    }

    public DriverTimeView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public DriverTimeView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }


    @Override
    public void setCurrentNanoTime(long currentNanoTime) {
        super.setCurrentNanoTime(currentNanoTime);
        if (session == null) return;
        setText(TimeUtils.formatNano(currentNanoTime - session.getStartDurationTime() + prevDuration));
    }

    public long getPrevDuration() {
        return prevDuration;
    }

    public void setPrevDuration(long prevDuration) {
        this.prevDuration = prevDuration;
    }

    public Session getSession() {
        return session;
    }

    public void setSession(Session session) {
        this.session = session;
        if (session != null && session.getStartDurationTime() == -1) setText(TimeUtils.format(0));
    }
}

File: app/src/main/java/ua/hospes/rtm/widgets/DriverTimeView.kt
package ua.hospes.rtm.widgets

import android.content.Context
import android.util.AttributeSet

import ua.hospes.rtm.domain.sessions.Session
import ua.hospes.rtm.utils.TimeUtils

class DriverTimeView : TimeView {
    var session: Session? = null
        set(session) {
            field = session
            val sessionDuration = with(session ?: return) {
                when {
                    startTime == null -> 0L
                    endTime != null -> endTime - startTime
                    else -> 0L
                }
            }

            text = TimeUtils.formatNano(sessionDuration + prevDuration)
        }
    var prevDuration = 0L
        set(prevDuration) {
            field = prevDuration
            val sessionDuration = with(session ?: return) {
                when {
                    startTime == null -> 0L
                    endTime != null -> endTime - startTime
                    else -> 0L
                }
            }

            text = TimeUtils.formatNano(sessionDuration + prevDuration)
        }


    override var currentNanoTime: Long
        get() = super.currentNanoTime
        set(currentNanoTime) {
            super.currentNanoTime = currentNanoTime
            with(session ?: return) {
                text = TimeUtils.formatNano(currentNanoTime - (startTime ?: 0L) + prevDuration)
            }
        }


    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)
}